File: packages/swt3-mcp/src/fingerprint.ts
/**
 * SWT3 MCP Server — Fingerprint minting and SHA-256 utilities.
 *
 * Inlined from @tenova/swt3-ai to keep this package self-contained.
 * The fingerprint formula is LOCKED:
 *   SHA256("WITNESS:{tenant}:{proc}:{fa}:{fb}:{fc}:{ts_ms}").hex().slice(0, 12)
 */

import { createHash, createHmac } from "crypto";

export function sha256Hex(data: string, length: number = 64): string {
  return createHash("sha256").update(data, "utf-8").digest("hex").slice(0, length);
}

export function sha256Truncated(data: string, length: number = 16): string {
  return sha256Hex(data, length);
}

function numStr(v: number): string {
  return String(v);
}

export function mintFingerprint(
  tenantId: string,
  procedureId: string,
  factorA: number,
  factorB: number,
  factorC: number,
  timestampMs: number,
): string {
  const fpInput =
    `WITNESS:${tenantId}:${procedureId}` +
    `:${numStr(factorA)}:${numStr(factorB)}:${numStr(factorC)}` +
    `:${timestampMs}`;
  return createHash("sha256").update(fpInput, "utf-8").digest("hex").slice(0, 12);
}

export function timestampMs(): [number, number] {
  const ts = Date.now();
  const epoch = Math.floor(ts / 1000);
  return [ts, epoch];
}

export function signPayload(
  signingKey: string,
  anchorFingerprint: string,
  agentId?: string,
): string {
  const message = agentId
    ? `${anchorFingerprint}:${agentId}`
    : anchorFingerprint;
  return createHmac("sha256", signingKey)
    .update(message, "utf-8")
    .digest("hex");
}

File: packages/swt3-ai-kotlin/src/main/kotlin/io/tenova/swt3/Fingerprint.kt
package io.tenova.swt3

import java.security.MessageDigest

/**
 * SWT3 fingerprint formula -- LOCKED, cross-language parity with 7 other SDKs.
 *
 * Formula: SHA256("WITNESS:{tenant}:{proc}:{fa}:{fb}:{fc}:{ts_ms}").hex()[:12]
 */
object Fingerprint {

    /**
     * Mint a 12-character hex fingerprint from witness factors.
     *
     * The fingerprint is deterministic: same inputs always produce the same output
     * across Python, TypeScript, Rust, C#, Ruby, Swift, and Kotlin.
     */
    fun mintFingerprint(
        tenant: String,
        procedure: String,
        factorA: Double,
        factorB: Double,
        factorC: Double,
        timestampMs: Long,
    ): String {
        val message = "WITNESS:$tenant:$procedure:${formatFactor(factorA)}:${formatFactor(factorB)}:${formatFactor(factorC)}:$timestampMs"
        return sha256Hex(message).substring(0, 12)
    }

    /**
     * SHA-256 hash truncated to [length] hex characters (default 16).
     */
    fun sha256Truncated(data: String, length: Int = 16): String {
        return sha256Hex(data).substring(0, length)
    }

    /**
     * Full SHA-256 hex digest (64 characters).
     */
    fun sha256Hex(data: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val bytes = digest.digest(data.toByteArray(Charsets.UTF_8))
        return bytes.joinToString("") { "%02x".format(it) }
    }

    /**
     * Current timestamp as (milliseconds, seconds) pair.
     */
    fun timestampMs(): Pair<Long, Long> {
        val ms = System.currentTimeMillis()
        return Pair(ms, ms / 1000)
    }

    /**
     * Format a numeric factor for the fingerprint formula.
     * Integer-valued doubles lose their decimal: 1.0 -> "1", 0.0 -> "0"
     * True floats keep decimals: 1.5 -> "1.5"
     */
    internal fun formatFactor(value: Double): String {
        return if (value == value.toLong().toDouble()) {
            value.toLong().toString()
        } else {
            value.toString()
        }
    }
}
